File: csharp/circular-buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;

public class CircularBuffer<T>
{
    private readonly int _capacity;
    private readonly Queue<T> _queue;
    
    public CircularBuffer(int capacity)
    {
        _capacity = capacity;
        _queue = new Queue<T>();
    }

    public T Read() => _queue.Dequeue();

    public void Write(T value)
    {
        if (_queue.Count == _capacity)
        {
            throw new InvalidOperationException();
        }
        
        _queue.Enqueue(value);
    }

    public void Overwrite(T value)
    {
        if (_queue.Count == _capacity)
        {
            _queue.Dequeue();
        }
        
        _queue.Enqueue(value);
    }

    public void Clear() => _queue.Clear();
}

File: swift/circular-buffer/Sources/CircularBuffer/CircularBuffer.swift
enum CircularBufferError: Error {
  case bufferEmpty
  case bufferFull
}

struct CircularBuffer<T> {
  private var items: [T]
  private let capacity: Int

  init(capacity: Int) {
    self.items = []
    self.capacity = capacity
  }

  mutating func read() throws -> T {
    guard !items.isEmpty else {
      throw CircularBufferError.bufferEmpty
    }

    return items.removeFirst()
  }

  mutating func write(_ item: T) throws {
    guard items.count < capacity else {
      throw CircularBufferError.bufferFull
    }

    items.append(item)
  }

  mutating func overwrite(_ item: T) {
    if items.count == capacity {
      _ = items.removeFirst()
    }

    items.append(item)
  }

  mutating func clear() {
    items.removeAll()
  }
}
